File: lab3/lab3-2.java
import java.util.HashSet;

public class Solution {
    public String longestNiceSubstring(String s) {
        if (s.length() < 2) {
            return "";
        }

        for (int i = 0; i < s.length(); i++) {
            if (!s.contains(String.valueOf(swapCase(s.charAt(i))))) {
                String left = longestNiceSubstring(s.substring(0, i));
                String right = longestNiceSubstring(s.substring(i + 1));
                return left.length() >= right.length() ? left : right;
            }
        }

        return s;
    }

    private char swapCase(char c) {
        if (Character.isLowerCase(c)) {
            return Character.toUpperCase(c);
        } else {
            return Character.toLowerCase(c);
        }
    }

    
}

File: lab3/lab3-2.py
class Solution(object):
    def longestNiceSubstring(self, s):
    
        def isNice(sub):
            seen = set(sub)
            for c in seen:
                if c.swapcase() not in seen:
                    return False
            return True
        
        if len(s) < 2:
            return ""

        for i in range(len(s)):
            if s[i].swapcase() not in s:
                left = self.longestNiceSubstring(s[:i])
                right = self.longestNiceSubstring(s[i + 1:])
                return left if len(left) >= len(right) else right

        return s


